packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -181,125 +194,274 @@ class CupertinoAlertDialog extends StatelessWidget {
+                    id: _AlertDialogSections.actionsSection,
+                    child: _buildActions(),
+                  ),
+                ],
+              ),
+            ),
+          ),
         ),
       ),
     );
   }
 }
 
-/// A button typically used in a [CupertinoAlertDialog].
-///
-/// See also:
-///
-///  * [CupertinoAlertDialog], a dialog that informs the user about situations
-///    that require acknowledgement
-class CupertinoDialogAction extends StatelessWidget {
-  /// Creates an action for an iOS-style dialog.
-  const CupertinoDialogAction({
-    this.onPressed,
-    this.isDefaultAction = false,
-    this.isDestructiveAction = false,
-    @required this.child,
-  }) : assert(child != null);
+// iOS style layout policy widget for sizing an alert dialog's content section and
+// action button section.
+//
+// See [_RenderCupertinoDialog] for specific layout policy details.
+class _CupertinoDialogRenderWidget extends MultiChildRenderObjectWidget {
+  _CupertinoDialogRenderWidget({
+    Key key,
+    @required List<Widget> children,
+  }) : super(key: key, children: children);
 
-  /// The callback that is called when the button is tapped or otherwise
-  /// activated.
-  ///
-  /// If this is set to null, the button will be disabled.
-  final VoidCallback onPressed;
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return new _RenderCupertinoDialog();
+  }
+}
 
-  /// Set to true if button is the default choice in the dialog.
-  ///
-  /// Default buttons are bold.
-  final bool isDefaultAction;
+// iOS style layout policy for sizing an alert dialog's content section and action
+// button section.
+//
+// The policy is as follows:
+//
+// If all content and buttons fit on screen:
+// The content section and action button section are sized intrinsically and centered
+// vertically on screen.
+//
+// If all content and buttons do not fit on screen:
+// A minimum height for the action button section is calculated. The action
+// button section will not be rendered shorter than this minimum.  See
+// [_RenderCupertinoDialogActions] for the minimum height calculation.
+//
+// With the minimum action button section calculated, the content section can
+// take up as much space as is available, up to the point that it hits the
+// minimum button height at the bottom.
+//
+// After the content section is laid out, the action button section is allowed
+// to take up any remaining space that was not consumed by the content section.
+class _RenderCupertinoDialog extends RenderBox
+    with ContainerRenderObjectMixin<RenderBox, MultiChildLayoutParentData>,
+    RenderBoxContainerDefaultsMixin<RenderBox, MultiChildLayoutParentData> {
+  _RenderCupertinoDialog({
+    RenderBox contentSection,
+    RenderBox actionsSection,
+  }) {
+    if (null != contentSection) {
+      add(contentSection);
+    }
+    if (null != actionsSection) {
+      add(actionsSection);
+    }
```

Comment:
if you just have two children, why store them in a list? that seems highly inefficient. Just have two fields, and implement the adopt/drop logic (see the mixin for single children, just do it twice).

Reply:
Ok, I introduced a custom RenderElement and now the Widget/Element/RenderBox all operate on 2 explicit children.